File: frontend/lib/features/videos/services/video_upload_service.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:dio/dio.dart';
import 'package:video_compress/video_compress.dart';

/// Video upload service with compression and chunked upload support
class VideoUploadService {
  final Dio _dio;
  final String _baseUrl;

  // Upload configuration
  static const int chunkSize = 1536 * 1024; // 1.5MB chunks (fits within PHP's 2MB default)
  static const int maxFileSize = 500 * 1024 * 1024; // 500MB max

  VideoUploadService({
    required Dio dio,
    required String baseUrl,
  })  : _dio = dio,
        _baseUrl = baseUrl;

  /// Compress video using video_compress package
  /// Returns the path to the compressed video
  Future<VideoCompressionResult> compressVideo(
    String inputPath, {
    void Function(double progress)? onProgress,
    VideoQuality quality = VideoQuality.MediumQuality,
  }) async {
    final inputFile = File(inputPath);
    if (!await inputFile.exists()) {
      throw VideoUploadException('Input file does not exist');
    }

    // Get original file info
    final originalSize = await inputFile.length();

    // Subscribe to compression progress
    final subscription = VideoCompress.compressProgress$.subscribe((progress) {
      onProgress?.call(progress / 100);
    });

    try {
      // Compress the video
      final info = await VideoCompress.compressVideo(
        inputPath,
        quality: quality,
        deleteOrigin: false,
        includeAudio: true,
      );

      if (info == null || info.path == null) {
        throw VideoUploadException('Video compression failed');
      }

      final compressedFile = File(info.path!);
      final compressedSize = await compressedFile.length();

      return VideoCompressionResult(
        originalPath: inputPath,
        compressedPath: info.path!,
        originalSize: originalSize,
        compressedSize: compressedSize,
        compressionRatio: originalSize / compressedSize,
        width: info.width ?? 0,
        height: info.height ?? 0,
        durationMs: (info.duration ?? 0).toInt(),
      );
    } finally {
      subscription.unsubscribe();
    }
  }

  /// Upload video using chunked upload
  Future<ChunkedUploadResult> uploadVideo(
    String filePath, {
    required String directory,
    required String authToken,
    String? title,
    String? description,
    int? businessId,
    bool createVideoRecord = false,
    void Function(double progress, String status)? onProgress,
    CancelToken? cancelToken,
  }) async {
    final file = File(filePath);
    if (!await file.exists()) {
      throw VideoUploadException('File does not exist');
    }

    final fileSize = await file.length();
    if (fileSize > maxFileSize) {
      throw VideoUploadException('File size exceeds maximum allowed size of ${maxFileSize ~/ (1024 * 1024)}MB');
    }

    final fileName = file.path.split(Platform.pathSeparator).last;
    final mimeType = _getMimeType(fileName);

    // Step 1: Initialize upload
    onProgress?.call(0.0, 'Initializing upload...');

    final initResponse = await _dio.post(
      '$_baseUrl/upload/chunked/init',
      data: {
        'file_name': fileName,
        'file_size': fileSize,
        'mime_type': mimeType,
        'chunk_size': chunkSize,
      },
      options: Options(
        headers: {'Authorization': 'Bearer $authToken'},
      ),
      cancelToken: cancelToken,
    );

    final uploadId = initResponse.data['data']['upload_id'] as String;
    final totalChunks = initResponse.data['data']['total_chunks'] as int;

    // Step 2: Upload chunks
    final fileStream = file.openRead();
    int chunkIndex = 0;

    await for (final chunk in _chunkStream(fileStream, chunkSize)) {
      if (cancelToken?.isCancelled ?? false) {
        await _cancelUpload(uploadId, authToken);
        throw VideoUploadException('Upload cancelled');
      }

      final progress = (chunkIndex + 1) / totalChunks;
      onProgress?.call(progress * 0.9, 'Uploading chunk ${chunkIndex + 1}/$totalChunks...');

      await _uploadChunk(
        uploadId: uploadId,
        chunkIndex: chunkIndex,
        chunkData: chunk,
        authToken: authToken,
        cancelToken: cancelToken,
      );

      chunkIndex++;
    }

    // Step 3: Complete upload
    onProgress?.call(0.95, 'Finalizing upload...');

    final completeResponse = await _dio.post(
      '$_baseUrl/upload/chunked/complete',
      data: {
        'upload_id': uploadId,
        'directory': directory,
        'create_video': createVideoRecord,
        if (businessId != null) 'business_id': businessId,
        if (title != null) 'title': title,
        if (description != null) 'description': description,
      },
      options: Options(
        headers: {'Authorization': 'Bearer $authToken'},
      ),
      cancelToken: cancelToken,
    );

    onProgress?.call(1.0, 'Upload complete!');

    return ChunkedUploadResult(
      fileUrl: completeResponse.data['data']['file_url'] as String,
      filePath: completeResponse.data['data']['file_path'] as String,
      fileName: completeResponse.data['data']['file_name'] as String,
      fileSize: completeResponse.data['data']['file_size'] as int,
      videoId: completeResponse.data['data']['video']?['id'] as int?,
      videoUuid: completeResponse.data['data']['video']?['uuid'] as String?,
    );
  }

  /// Compress and upload video in one operation
  Future<ChunkedUploadResult> compressAndUpload(
    String inputPath, {
    required String directory,
    required String authToken,
    String? title,
    String? description,
    int? businessId,
    bool createVideoRecord = false,
    void Function(double progress, String status)? onProgress,
    CancelToken? cancelToken,
  }) async {
    // Step 1: Compress video (0-50% of progress)
    onProgress?.call(0.0, 'Compressing video...');

    final compressionResult = await compressVideo(
      inputPath,
      onProgress: (progress) {
        onProgress?.call(progress * 0.5, 'Compressing video... ${(progress * 100).toInt()}%');
      },
    );

    // Step 2: Upload compressed video (50-100% of progress)
    try {
      final result = await uploadVideo(
        compressionResult.compressedPath,
        directory: directory,
        authToken: authToken,
        title: title,
        description: description,
        businessId: businessId,
        createVideoRecord: createVideoRecord,
        onProgress: (progress, status) {
          onProgress?.call(0.5 + (progress * 0.5), status);
        },
        cancelToken: cancelToken,
      );

      return result;
    } finally {
      // Clean up compressed file
      final compressedFile = File(compressionResult.compressedPath);
      if (await compressedFile.exists()) {
        await compressedFile.delete();
      }
      // Cancel any ongoing compression
      await VideoCompress.cancelCompression();
    }
  }

  /// Get upload status for resuming
  Future<UploadStatus> getUploadStatus(String uploadId, String authToken) async {
    final response = await _dio.get(
      '$_baseUrl/upload/chunked/status/$uploadId',
      options: Options(
        headers: {'Authorization': 'Bearer $authToken'},
      ),
    );

    final data = response.data['data'];
    return UploadStatus(
      uploadId: data['upload_id'],
      fileName: data['file_name'],
      totalSize: data['total_size'],
      totalChunks: data['total_chunks'],
      uploadedChunks: List<int>.from(data['uploaded_chunks']),
      progress: data['progress'].toDouble(),
      isComplete: data['is_complete'],
    );
  }

  /// Resume a failed upload
  Future<ChunkedUploadResult> resumeUpload(
    String filePath,
    String uploadId, {
    required String directory,
    required String authToken,
    String? title,
    String? description,
    int? businessId,
    bool createVideoRecord = false,
    void Function(double progress, String status)? onProgress,
    CancelToken? cancelToken,
  }) async {
    // Get current status
    final status = await getUploadStatus(uploadId, authToken);

    if (status.isComplete) {
      throw VideoUploadException('Upload is already complete');
    }

    final file = File(filePath);
    final missingChunks = List<int>.generate(status.totalChunks, (i) => i)
        .where((i) => !status.uploadedChunks.contains(i))
        .toList();

    // Upload missing chunks
    for (var i = 0; i < missingChunks.length; i++) {
      final chunkIndex = missingChunks[i];
      final offset = chunkIndex * chunkSize;

      if (cancelToken?.isCancelled ?? false) {
        throw VideoUploadException('Upload cancelled');
      }

      final progress = (status.uploadedChunks.length + i + 1) / status.totalChunks;
      onProgress?.call(progress * 0.9, 'Resuming upload... chunk ${chunkIndex + 1}/${status.totalChunks}');

      final raf = await file.open();
      await raf.setPosition(offset);
      final chunkData = await raf.read(chunkSize);
      await raf.close();

      await _uploadChunk(
        uploadId: uploadId,
        chunkIndex: chunkIndex,
        chunkData: chunkData,
        authToken: authToken,
        cancelToken: cancelToken,
      );
    }

    // Complete upload
    onProgress?.call(0.95, 'Finalizing upload...');

    final completeResponse = await _dio.post(
      '$_baseUrl/upload/chunked/complete',
      data: {
        'upload_id': uploadId,
        'directory': directory,
        'create_video': createVideoRecord,
        if (businessId != null) 'business_id': businessId,
        if (title != null) 'title': title,
        if (description != null) 'description': description,
      },
      options: Options(
        headers: {'Authorization': 'Bearer $authToken'},
      ),
      cancelToken: cancelToken,
    );

    onProgress?.call(1.0, 'Upload complete!');

    return ChunkedUploadResult(
      fileUrl: completeResponse.data['data']['file_url'] as String,
      filePath: completeResponse.data['data']['file_path'] as String,
      fileName: completeResponse.data['data']['file_name'] as String,
      fileSize: completeResponse.data['data']['file_size'] as int,
      videoId: completeResponse.data['data']['video']?['id'] as int?,
      videoUuid: completeResponse.data['data']['video']?['uuid'] as String?,
    );
  }

  // Private helper methods

  Future<void> _uploadChunk({
    required String uploadId,
    required int chunkIndex,
    required Uint8List chunkData,
    required String authToken,
    CancelToken? cancelToken,
  }) async {
    final formData = FormData.fromMap({
      'upload_id': uploadId,
      'chunk_index': chunkIndex,
      'chunk': MultipartFile.fromBytes(
        chunkData,
        filename: 'chunk_$chunkIndex',
      ),
    });

    await _dio.post(
      '$_baseUrl/upload/chunked/chunk',
      data: formData,
      options: Options(
        headers: {'Authorization': 'Bearer $authToken'},
      ),
      cancelToken: cancelToken,
    );
  }

  Future<void> _cancelUpload(String uploadId, String authToken) async {
    try {
      await _dio.delete(
        '$_baseUrl/upload/chunked/cancel/$uploadId',
        options: Options(
          headers: {'Authorization': 'Bearer $authToken'},
        ),
      );
    } catch (_) {
      // Ignore cancel errors
    }
  }

  Stream<Uint8List> _chunkStream(Stream<List<int>> source, int chunkSize) async* {
    var buffer = <int>[];

    await for (final data in source) {
      buffer.addAll(data);

      while (buffer.length >= chunkSize) {
        yield Uint8List.fromList(buffer.sublist(0, chunkSize));
        buffer = buffer.sublist(chunkSize);
      }
    }

    if (buffer.isNotEmpty) {
      yield Uint8List.fromList(buffer);
    }
  }

  String _getMimeType(String fileName) {
    final ext = fileName.split('.').last.toLowerCase();
    switch (ext) {
      case 'mp4':
        return 'video/mp4';
      case 'mov':
        return 'video/quicktime';
      case 'avi':
        return 'video/x-msvideo';
      case 'webm':
        return 'video/webm';
      case 'mpeg':
      case 'mpg':
        return 'video/mpeg';
      default:
        return 'video/mp4';
    }
  }
}

// Data classes

class VideoCompressionResult {
  final String originalPath;
  final String compressedPath;
  final int originalSize;
  final int compressedSize;
  final double compressionRatio;
  final int width;
  final int height;
  final int durationMs;

  VideoCompressionResult({
    required this.originalPath,
    required this.compressedPath,
    required this.originalSize,
    required this.compressedSize,
    required this.compressionRatio,
    required this.width,
    required this.height,
    required this.durationMs,
  });

  String get savedPercentage => '${((1 - (compressedSize / originalSize)) * 100).toStringAsFixed(1)}%';
}

class ChunkedUploadResult {
  final String fileUrl;
  final String filePath;
  final String fileName;
  final int fileSize;
  final int? videoId;
  final String? videoUuid;

  ChunkedUploadResult({
    required this.fileUrl,
    required this.filePath,
    required this.fileName,
    required this.fileSize,
    this.videoId,
    this.videoUuid,
  });
}

class UploadStatus {
  final String uploadId;
  final String fileName;
  final int totalSize;
  final int totalChunks;
  final List<int> uploadedChunks;
  final double progress;
  final bool isComplete;

  UploadStatus({
    required this.uploadId,
    required this.fileName,
    required this.totalSize,
    required this.totalChunks,
    required this.uploadedChunks,
    required this.progress,
    required this.isComplete,
  });
}

class VideoUploadException implements Exception {
  final String message;
  VideoUploadException(this.message);

  @override
  String toString() => 'VideoUploadException: $message';
}

File: resources/js/services/videoUploadService.js
import { FFmpeg } from '@ffmpeg/ffmpeg';
import { fetchFile, toBlobURL } from '@ffmpeg/util';

// FFmpeg core files are in public/ffmpeg/ folder (loaded at runtime, not imported)
const FFMPEG_BASE_URL = '/ffmpeg';

const CHUNK_SIZE = 1.5 * 1024 * 1024; // 1.5MB chunks (fits within PHP's 2MB default limit)
const MAX_FILE_SIZE = 500 * 1024 * 1024; // 500MB max

class VideoUploadService {
    constructor() {
        this.ffmpeg = null;
        this.ffmpegLoaded = false;
        this.ffmpegLoading = false;
    }

    /**
     * Load FFmpeg.wasm for browser-side video compression
     */
    async loadFFmpeg(onProgress) {
        if (this.ffmpegLoaded) return;
        if (this.ffmpegLoading) {
            // Wait for existing load to complete
            while (this.ffmpegLoading) {
                await new Promise(resolve => setTimeout(resolve, 100));
            }
            return;
        }

        this.ffmpegLoading = true;

        try {
            this.ffmpeg = new FFmpeg();

            this.ffmpeg.on('log', ({ message }) => {
                console.log('[FFmpeg]', message);
            });

            this.ffmpeg.on('progress', ({ progress }) => {
                if (onProgress) {
                    onProgress(progress);
                }
            });

            // Load FFmpeg core from local public/ffmpeg folder
            // toBlobURL fetches the file and creates a blob URL (avoids import issues)
            const coreURL = await toBlobURL(`${FFMPEG_BASE_URL}/ffmpeg-core.js`, 'text/javascript');
            const wasmURL = await toBlobURL(`${FFMPEG_BASE_URL}/ffmpeg-core.wasm`, 'application/wasm');

            await this.ffmpeg.load({
                coreURL,
                wasmURL,
            });

            this.ffmpegLoaded = true;
        } catch (error) {
            console.error('Failed to load FFmpeg:', error);
            throw new Error('Failed to initialize video compression. Please try again.');
        } finally {
            this.ffmpegLoading = false;
        }
    }

    /**
     * Compress video using FFmpeg.wasm in the browser
     */
    async compressVideo(file, onProgress) {
        if (!this.ffmpegLoaded) {
            await this.loadFFmpeg();
        }

        const inputName = 'input' + this.getExtension(file.name);
        const outputName = 'output.mp4';

        try {
            // Write input file to FFmpeg virtual filesystem
            await this.ffmpeg.writeFile(inputName, await fetchFile(file));

            // Compress video with optimal settings for web
            // - CRF 28: Good balance of quality and size
            // - preset: medium for reasonable encoding speed
            // - scale: max 1280 width while maintaining aspect ratio
            // - audio: AAC 128k
            await this.ffmpeg.exec([
                '-i', inputName,
                '-c:v', 'libx264',
                '-crf', '28',
                '-preset', 'medium',
                '-vf', 'scale=min(1280\\,iw):-2',
                '-c:a', 'aac',
                '-b:a', '128k',
                '-movflags', '+faststart',
                '-y', outputName
            ]);

            // Read compressed file
            const data = await this.ffmpeg.readFile(outputName);

            // Cleanup
            await this.ffmpeg.deleteFile(inputName);
            await this.ffmpeg.deleteFile(outputName);

            // Create blob from compressed data
            const blob = new Blob([data.buffer], { type: 'video/mp4' });

            return {
                file: new File([blob], 'compressed_video.mp4', { type: 'video/mp4' }),
                originalSize: file.size,
                compressedSize: blob.size,
                compressionRatio: (file.size / blob.size).toFixed(2),
                savedPercentage: ((1 - blob.size / file.size) * 100).toFixed(1)
            };
        } catch (error) {
            console.error('Video compression failed:', error);
            throw new Error('Video compression failed. Uploading original file instead.');
        }
    }

    /**
     * Initialize chunked upload session
     */
    async initializeUpload(file, api) {
        if (file.size > MAX_FILE_SIZE) {
            throw new Error(`File size exceeds maximum allowed size of ${MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }

        const response = await api.post('/upload/chunked/init', {
            file_name: file.name,
            file_size: file.size,
            mime_type: file.type || 'video/mp4',
            chunk_size: CHUNK_SIZE
        });

        return response.data.data;
    }

    /**
     * Upload a single chunk
     */
    async uploadChunk(uploadId, chunkIndex, chunk, api) {
        const formData = new FormData();
        formData.append('upload_id', uploadId);
        formData.append('chunk_index', chunkIndex.toString());
        // Create a proper Blob with video mime type
        const chunkBlob = new Blob([chunk], { type: 'application/octet-stream' });
        formData.append('chunk', chunkBlob, `chunk_${chunkIndex}.bin`);

        try {
            const response = await api.post('/upload/chunked/chunk', formData, {
                headers: { 'Content-Type': 'multipart/form-data' }
            });
            return response.data.data;
        } catch (error) {
            console.error('Chunk upload failed:', {
                chunkIndex,
                chunkSize: chunk.size || chunk.byteLength,
                error: error.response?.data || error.message
            });
            throw error;
        }
    }

    /**
     * Complete chunked upload
     */
    async completeUpload(uploadId, directory, api, options = {}) {
        const response = await api.post('/upload/chunked/complete', {
            upload_id: uploadId,
            directory: directory,
            create_video: options.createVideo || false,
            business_id: options.businessId,
            title: options.title,
            description: options.description
        });

        return response.data.data;
    }

    /**
     * Get upload status for resume capability
     */
    async getUploadStatus(uploadId, api) {
        const response = await api.get(`/upload/chunked/status/${uploadId}`);
        return response.data.data;
    }

    /**
     * Cancel an upload
     */
    async cancelUpload(uploadId, api) {
        try {
            await api.delete(`/upload/chunked/cancel/${uploadId}`);
        } catch (error) {
            console.error('Failed to cancel upload:', error);
        }
    }

    /**
     * Upload video with compression and chunked upload
     */
    async uploadVideo(file, api, options = {}) {
        const {
            directory = 'videos',
            onProgress,
            onStatusChange,
            compress = true,
            createVideo = false,
            businessId,
            title,
            description,
            abortController
        } = options;

        let videoFile = file;
        let compressionResult = null;

        // Step 1: Compress video (if enabled and FFmpeg is available)
        if (compress) {
            try {
                onStatusChange?.('Initializing compression...');
                await this.loadFFmpeg((progress) => {
                    onProgress?.(progress * 0.1); // 0-10% for loading FFmpeg
                    onStatusChange?.('Loading compression engine...');
                });

                onStatusChange?.('Compressing video...');
                compressionResult = await this.compressVideo(file, (progress) => {
                    onProgress?.(0.1 + progress * 0.4); // 10-50% for compression
                    onStatusChange?.(`Compressing video... ${Math.round(progress * 100)}%`);
                });

                videoFile = compressionResult.file;
                onStatusChange?.(`Compression complete! Saved ${compressionResult.savedPercentage}%`);
            } catch (error) {
                console.warn('Compression failed, uploading original:', error);
                onStatusChange?.('Compression unavailable, uploading original...');
            }
        }

        // Step 2: Initialize chunked upload
        onStatusChange?.('Initializing upload...');
        const uploadSession = await this.initializeUpload(videoFile, api);
        const { upload_id: uploadId, total_chunks: totalChunks, chunk_size: chunkSize } = uploadSession;

        // Step 3: Upload chunks
        const baseProgress = compress ? 0.5 : 0; // Start at 50% if compressed
        const progressRange = compress ? 0.45 : 0.95; // 50-95% if compressed, 0-95% if not

        for (let chunkIndex = 0; chunkIndex < totalChunks; chunkIndex++) {
            // Check if upload was cancelled
            if (abortController?.signal?.aborted) {
                await this.cancelUpload(uploadId, api);
                throw new Error('Upload cancelled');
            }

            const start = chunkIndex * chunkSize;
            const end = Math.min(start + chunkSize, videoFile.size);
            const chunk = videoFile.slice(start, end);

            const chunkProgress = (chunkIndex + 1) / totalChunks;
            onProgress?.(baseProgress + chunkProgress * progressRange);
            onStatusChange?.(`Uploading chunk ${chunkIndex + 1}/${totalChunks}...`);

            await this.uploadChunk(uploadId, chunkIndex, chunk, api);
        }

        // Step 4: Complete upload
        onProgress?.(0.97);
        onStatusChange?.('Finalizing upload...');

        const result = await this.completeUpload(uploadId, directory, api, {
            createVideo,
            businessId,
            title,
            description
        });

        onProgress?.(1);
        onStatusChange?.('Upload complete!');

        return {
            ...result,
            compressionResult
        };
    }

    /**
     * Resume a failed upload
     */
    async resumeUpload(file, uploadId, api, options = {}) {
        const {
            directory = 'videos',
            onProgress,
            onStatusChange,
            createVideo = false,
            businessId,
            title,
            description
        } = options;

        // Get current upload status
        const status = await this.getUploadStatus(uploadId, api);

        if (status.is_complete) {
            throw new Error('Upload is already complete');
        }

        // Find missing chunks
        const allChunks = Array.from({ length: status.total_chunks }, (_, i) => i);
        const missingChunks = allChunks.filter(i => !status.uploaded_chunks.includes(i));

        onStatusChange?.(`Resuming upload... ${missingChunks.length} chunks remaining`);

        // Upload missing chunks
        for (let i = 0; i < missingChunks.length; i++) {
            const chunkIndex = missingChunks[i];
            const start = chunkIndex * CHUNK_SIZE;
            const end = Math.min(start + CHUNK_SIZE, file.size);
            const chunk = file.slice(start, end);

            const progress = (status.uploaded_chunks.length + i + 1) / status.total_chunks;
            onProgress?.(progress * 0.95);
            onStatusChange?.(`Uploading chunk ${chunkIndex + 1}/${status.total_chunks}...`);

            await this.uploadChunk(uploadId, chunkIndex, chunk, api);
        }

        // Complete upload
        onProgress?.(0.97);
        onStatusChange?.('Finalizing upload...');

        const result = await this.completeUpload(uploadId, directory, api, {
            createVideo,
            businessId,
            title,
            description
        });

        onProgress?.(1);
        onStatusChange?.('Upload complete!');

        return result;
    }

    /**
     * Get file extension
     */
    getExtension(filename) {
        const ext = filename.split('.').pop().toLowerCase();
        return ext ? `.${ext}` : '.mp4';
    }

    /**
     * Check if FFmpeg.wasm is supported in this browser
     */
    isCompressionSupported() {
        return typeof SharedArrayBuffer !== 'undefined';
    }
}

// Export singleton instance
export const videoUploadService = new VideoUploadService();

// Export class for testing
export default VideoUploadService;
